// @lark-opdev/block-docs-addon-webpack-utils/dist/es/index.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/libs/webpack_plugin.js
import { Compilation, sources } from "webpack";
import { cyan } from "colorette";
import open from "open";
import qs from "query-string";

// src/libs/utils.js
import path from "path";
import opapi from "@lark-opdev/cli/libs/api";
import axios from "axios";
import fs from "fs";
import merge from "lodash.merge";
function getAccountInfo() {
  return __async(this, null, function* () {
    const config = JSON.parse(fs.readFileSync(path.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    let api;
    if (config && config.privatization) {
      api = yield opapi.init({
        envConfig: config.privatization
      });
    } else {
      api = yield opapi.init();
    }
    const account = api.account.getAccountInfo();
    if (!account || !account.larkSession || !account.env) {
      throw new Error("get lark session failed. please run opdev login before developing");
    }
    return account;
  });
}
function setDefaultOpenUrl(url) {
  const blockConfigPath = path.resolve(process.cwd(), "app.json");
  const config = JSON.parse(fs.readFileSync(blockConfigPath, { encoding: "utf-8" }));
  config.url = url;
  fs.writeFileSync(blockConfigPath, JSON.stringify(config, null, 2), { encoding: "utf-8" });
}
function getConfigFiles() {
  return __async(this, null, function* () {
    var _a;
    const env = process.env.CI_FLAG ? "feishu" /* feishu */ : (_a = yield getAccountInfo()) == null ? void 0 : _a.env;
    const config = JSON.parse(fs.readFileSync(path.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    const newConfig = merge(config, (config.environments || {})[env] || {});
    return {
      projectInfo: {
        appid: newConfig.appID,
        projectname: newConfig.projectName,
        blocks: ["index"]
      },
      blockInfo: {
        blockTypeID: newConfig.blockTypeID,
        blockRenderType: "offlineWeb",
        offlineWebConfig: {
          initialHeight: newConfig.initialHeight,
          contributes: newConfig.contributes
        }
      },
      docsaddon: {
        url: newConfig.url
      },
      privatization: newConfig.privatization
    };
  });
}
function createBaseFromTemplate(name) {
  return __async(this, null, function* () {
    const { larkSession, env } = yield getAccountInfo();
    const domain = `${env}.cn`;
    const instance = axios.create({
      baseURL: `https://internal-api-space.${domain}`,
      timeout: 5e3,
      headers: {
        "content-type": "application/x-www-form-urlencoded",
        cookie: `session=${larkSession}`,
        Referer: `https://${domain}`
      }
    });
    const res = yield instance.post("/space/api/explorer/v2/create/object/", {
      type: 22,
      name,
      source: 0
    });
    return Object.values(res.data.data.entities.nodes)[0].url;
  });
}
function getAppCSPConfig() {
  return __async(this, null, function* () {
    const api = yield opapi.init();
    return yield api.dev.getAppCSPConfig();
  });
}

// src/libs/webpack_plugin.js
var PLUGIN_NAME = "docverse-debug-url";
var DocVerseWebpackPlugin = class {
  constructor(options = {}) {
    __publicField(this, "url");
    __publicField(this, "open");
    __publicField(this, "isInit", false);
    __publicField(this, "port", 8080);
    this.open = typeof options.open !== "undefined" ? Boolean(options.open) : true;
    this.url = options.url;
  }
  apply(compiler) {
    if (process.env.NODE_ENV === "production") {
      compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
        compilation.hooks.processAssets.tapPromise(
          {
            name: "DocVerseWebpackPlugin",
            stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
          },
          () => __async(this, null, function* () {
            const { projectInfo, blockInfo } = yield getConfigFiles();
            compilation.emitAsset("project.config.json", new sources.RawSource(JSON.stringify(projectInfo)));
            compilation.emitAsset("index.json", new sources.RawSource(JSON.stringify(blockInfo)));
          })
        );
      });
    } else {
      compiler.hooks.infrastructureLog.tap(PLUGIN_NAME, (name, _type, args) => {
        if (name === "webpack-dev-server") {
          if (args[0].startsWith("Loopback:")) {
            this.port = args[0].match(/localhost:(\d+)/)[1];
          }
        }
      });
      compiler.hooks.done.tapPromise(PLUGIN_NAME, () => __async(this, null, function* () {
        if (this.isInit === false) {
          this.isInit = true;
          const {
            docsaddon: { url },
            privatization
          } = yield getConfigFiles();
          this.url = url != null ? url : "";
          if (!this.url) {
            if (privatization) {
              console.error(
                'Create docsaddon failure. Please set url of docsaddon to "url" property in app.json'
              );
              process.exit(1);
            }
            try {
              const newBaseUrl = yield createBaseFromTemplate("云文档小应用测试页面");
              this.url = newBaseUrl;
              console.info(
                `Create docsaddon successfully. The url ${newBaseUrl} will be written to app.json`
              );
              setDefaultOpenUrl(newBaseUrl);
            } catch (e) {
              console.error(
                'Create docsaddon failure. Please set url of docsaddon to "url" property in app.json'
              );
              process.exit(1);
            }
          }
          const logger = compiler.getInfrastructureLogger(PLUGIN_NAME);
          const url2 = qs.stringifyUrl({
            url: this.url,
            query: {
              blockitdebug: true,
              debugport: this.port
            }
          });
          logger.info(`URL: ${cyan(url2)}`);
          this.open && open(url2);
        }
      }));
    }
    return;
  }
};

// src/libs/dev_middleware.js
import opapi2 from "@lark-opdev/cli/libs/api";
import cors from "cors";
import path2 from "path";
import fs2 from "fs";
function dev_middleware_default(devServer) {
  return __async(this, null, function* () {
    devServer.app.use(
      cors({
        origin: true,
        credentials: true
      })
    );
    const { projectInfo, blockInfo } = yield getConfigFiles();
    const config = JSON.parse(fs2.readFileSync(path2.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    let api;
    if (config && config.privatization) {
      api = yield opapi2.init({
        envConfig: config.privatization
      });
    } else {
      api = yield opapi2.init();
    }
    const { port } = devServer.server.address();
    return api.dev.getBlockitDevMiddleware({
      projectConfig: projectInfo,
      blockConfigMap: { [blockInfo.blockTypeID]: blockInfo },
      devServerHost: `http://localhost:${port}`,
      basePath: "/block"
    });
  });
}
export {
  dev_middleware_default as docsAddonDevMiddleware,
  DocVerseWebpackPlugin as docsAddonWebpackPlugin,
  getAppCSPConfig,
  getConfigFiles
};
