// @lark-opdev/block-docs-addon-webpack-utils/dist/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.js
var src_exports = {};
__export(src_exports, {
  docsAddonDevMiddleware: () => dev_middleware_default,
  docsAddonWebpackPlugin: () => DocVerseWebpackPlugin,
  getAppCSPConfig: () => getAppCSPConfig,
  getConfigFiles: () => getConfigFiles
});
module.exports = __toCommonJS(src_exports);

// src/libs/webpack_plugin.js
var import_webpack = require("webpack");
var import_colorette = require("colorette");
var import_open = __toESM(require("open"));
var import_query_string = __toESM(require("query-string"));

// src/libs/utils.js
var import_path = __toESM(require("path"));
var import_api = __toESM(require("@lark-opdev/cli/libs/api"));
var import_axios = __toESM(require("axios"));
var import_fs = __toESM(require("fs"));
var import_lodash = __toESM(require("lodash.merge"));
function getAccountInfo() {
  return __async(this, null, function* () {
    const config = JSON.parse(import_fs.default.readFileSync(import_path.default.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    let api;
    if (config && config.privatization) {
      api = yield import_api.default.init({
        envConfig: config.privatization
      });
    } else {
      api = yield import_api.default.init();
    }
    const account = api.account.getAccountInfo();
    if (!account || !account.larkSession || !account.env) {
      throw new Error("get lark session failed. please run opdev login before developing");
    }
    return account;
  });
}
function setDefaultOpenUrl(url) {
  const blockConfigPath = import_path.default.resolve(process.cwd(), "app.json");
  const config = JSON.parse(import_fs.default.readFileSync(blockConfigPath, { encoding: "utf-8" }));
  config.url = url;
  import_fs.default.writeFileSync(blockConfigPath, JSON.stringify(config, null, 2), { encoding: "utf-8" });
}
function getConfigFiles() {
  return __async(this, null, function* () {
    var _a;
    const env = process.env.CI_FLAG ? "feishu" /* feishu */ : (_a = yield getAccountInfo()) == null ? void 0 : _a.env;
    const config = JSON.parse(import_fs.default.readFileSync(import_path.default.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    const newConfig = (0, import_lodash.default)(config, (config.environments || {})[env] || {});
    return {
      projectInfo: {
        appid: newConfig.appID,
        projectname: newConfig.projectName,
        blocks: ["index"]
      },
      blockInfo: {
        blockTypeID: newConfig.blockTypeID,
        blockRenderType: "offlineWeb",
        offlineWebConfig: {
          initialHeight: newConfig.initialHeight,
          contributes: newConfig.contributes
        }
      },
      docsaddon: {
        url: newConfig.url
      },
      privatization: newConfig.privatization
    };
  });
}
function createBaseFromTemplate(name) {
  return __async(this, null, function* () {
    const { larkSession, env } = yield getAccountInfo();
    const domain = `${env}.cn`;
    const instance = import_axios.default.create({
      baseURL: `https://internal-api-space.${domain}`,
      timeout: 5e3,
      headers: {
        "content-type": "application/x-www-form-urlencoded",
        cookie: `session=${larkSession}`,
        Referer: `https://${domain}`
      }
    });
    const res = yield instance.post("/space/api/explorer/v2/create/object/", {
      type: 22,
      name,
      source: 0
    });
    return Object.values(res.data.data.entities.nodes)[0].url;
  });
}
function getAppCSPConfig() {
  return __async(this, null, function* () {
    const api = yield import_api.default.init();
    return yield api.dev.getAppCSPConfig();
  });
}

// src/libs/webpack_plugin.js
var PLUGIN_NAME = "docverse-debug-url";
var DocVerseWebpackPlugin = class {
  constructor(options = {}) {
    __publicField(this, "url");
    __publicField(this, "open");
    __publicField(this, "isInit", false);
    __publicField(this, "port", 8080);
    this.open = typeof options.open !== "undefined" ? Boolean(options.open) : true;
    this.url = options.url;
  }
  apply(compiler) {
    if (process.env.NODE_ENV === "production") {
      compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
        compilation.hooks.processAssets.tapPromise(
          {
            name: "DocVerseWebpackPlugin",
            stage: import_webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
          },
          () => __async(this, null, function* () {
            const { projectInfo, blockInfo } = yield getConfigFiles();
            compilation.emitAsset("project.config.json", new import_webpack.sources.RawSource(JSON.stringify(projectInfo)));
            compilation.emitAsset("index.json", new import_webpack.sources.RawSource(JSON.stringify(blockInfo)));
          })
        );
      });
    } else {
      compiler.hooks.infrastructureLog.tap(PLUGIN_NAME, (name, _type, args) => {
        if (name === "webpack-dev-server") {
          if (args[0].startsWith("Loopback:")) {
            this.port = args[0].match(/localhost:(\d+)/)[1];
          }
        }
      });
      compiler.hooks.done.tapPromise(PLUGIN_NAME, () => __async(this, null, function* () {
        if (this.isInit === false) {
          this.isInit = true;
          const {
            docsaddon: { url },
            privatization
          } = yield getConfigFiles();
          this.url = url != null ? url : "";
          if (!this.url) {
            if (privatization) {
              console.error(
                'Create docsaddon failure. Please set url of docsaddon to "url" property in app.json'
              );
              process.exit(1);
            }
            try {
              const newBaseUrl = yield createBaseFromTemplate("云文档小应用测试页面");
              this.url = newBaseUrl;
              console.info(
                `Create docsaddon successfully. The url ${newBaseUrl} will be written to app.json`
              );
              setDefaultOpenUrl(newBaseUrl);
            } catch (e) {
              console.error(
                'Create docsaddon failure. Please set url of docsaddon to "url" property in app.json'
              );
              process.exit(1);
            }
          }
          const logger = compiler.getInfrastructureLogger(PLUGIN_NAME);
          const url2 = import_query_string.default.stringifyUrl({
            url: this.url,
            query: {
              blockitdebug: true,
              debugport: this.port
            }
          });
          logger.info(`URL: ${(0, import_colorette.cyan)(url2)}`);
          this.open && (0, import_open.default)(url2);
        }
      }));
    }
    return;
  }
};

// src/libs/dev_middleware.js
var import_api2 = __toESM(require("@lark-opdev/cli/libs/api"));
var import_cors = __toESM(require("cors"));
var import_path2 = __toESM(require("path"));
var import_fs2 = __toESM(require("fs"));
function dev_middleware_default(devServer) {
  return __async(this, null, function* () {
    devServer.app.use(
      (0, import_cors.default)({
        origin: true,
        credentials: true
      })
    );
    const { projectInfo, blockInfo } = yield getConfigFiles();
    const config = JSON.parse(import_fs2.default.readFileSync(import_path2.default.resolve(process.cwd(), "app.json"), { encoding: "utf-8" }));
    let api;
    if (config && config.privatization) {
      api = yield import_api2.default.init({
        envConfig: config.privatization
      });
    } else {
      api = yield import_api2.default.init();
    }
    const { port } = devServer.server.address();
    return api.dev.getBlockitDevMiddleware({
      projectConfig: projectInfo,
      blockConfigMap: { [blockInfo.blockTypeID]: blockInfo },
      devServerHost: `http://localhost:${port}`,
      basePath: "/block"
    });
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  docsAddonDevMiddleware,
  docsAddonWebpackPlugin,
  getAppCSPConfig,
  getConfigFiles
});
